src/pybind/mgr/telemetry/module.py

```diff
@@ -327,6 +331,110 @@ def gather_crashinfo(self) -> List[Dict[str, str]]:
             crashlist.append(c)
         return crashlist
 
+    def gather_perf_counters(self) -> Dict[str, dict]:
```

Comment:
This can simplify the process of generating the result dictionary:
```python
    import numbers

    def gather_perf_counters(self) -> Dict[str, dict]:
        all_perf_counters = self.get_all_perf_counters()
        result: Dict[str, dict] = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: defaultdict(int))))                                                    
        for daemon in all_perf_counters:
            daemon_type = daemon[0:3] # i.e. 'mds', 'osd', 'rgw'
            for collection in all_perf_counters[daemon]:
                col_0, col_1 = collection.split('.')
                if isinstance(all_perf_counters[daemon][collection]['value'], numbers.Number):
                    result[daemon_type][col_0][col_1]['value'] += all_perf_counters[daemon][collection]['value']                                                     
                if 'count' in all_perf_counters[daemon][collection]:
                    result[daemon_type][col_0][col_1]['count'] += all_perf_counters[daemon][collection]['count'] 
        return result
```

It leaves a 'value' key for each metric, and a 'count' key in case applicable. This keeps it consistent with the original names of the keys, and makes it easier on the server side too.
When we aggregate counters of similar daemon types, it's better if we send the raw sum (aggregated 'values'), and raw counts (aggregated 'count') to the telemetry server; most likely it will average the metrics between reports in order to draw a graph (we have the report timestamp, thus no need to average the time here).

If we wish to report (in the future) perf metrics per daemon (min/max/histogram per certain time granularity) we can have similar approach of sending the least processed data to the telemetry server (this way we don't lose information).

btw, some collections don't have a name, and just start with a '.' (like '.cache_bytes' in monitor's perf counters), which leads to an empty key in the result dictionary. This should probably be fixed on the C level (unless the collection does not have a name intentionally).

Reply:
All of this makes sense. I'll look in the C++ implementation to see where the nameless collections are happening.